### ISSUE TITLE
Cannot expand conditionals

### ISSUE THREAD

**Issue by shicks** [shicks/inlinedef#1]
@sasozivanovic

I'd like to conditionally define parts of a macro depending on whether the debug mode is on or not. So, assuming `\newif\ifdebug`, I'd like to write

```
\Inline\def\macro{%
   ...
   \ExpandIfT{debug}{debug code }%
   ...
}
```

With inlinedef, I'm almost there. I can define `\ExpandIfT` (and its cousins `\ExpandIfF` and `\ExpandIfTF` not shown here), and then expand them manually using `\MultiExpand`:

```
\def\ExpandIfT#1{%
   \csname
     \csname if#1\endcsname
       @firstofone%
     \else
       @gobble%
     \fi
   \endcsname
}

\newif\ifdebug
\debugtrue

\Inline\def\macro{%
   ...
   \MultiExpand{3}{\ExpandIfT{debug}{debug code }}%
   ...
}
```

However, being able to use `\ExpandIf(T)(F)` directly has quite an appeal to me. I investigated inlinedef code a bit and it doesn't seem very hard to extend it to accept user defined keywords inside an `\Inline` definition. I have drafted a patch makes this possible.

The idea is to define user commands in terms of `\Expand`, or, more probably, `\MultiExpand`. When a user keyword is encountered in the definition, it is expanded by the `\UnsafeExpand` mechanism.

I hope you will consider accepting the extension into the package.

**Comment by shicks** [comment]
@sasozivanovic also writes

On 03. 05. 2016 02:24, Stephen Hicks wrote:

> To be honest, I haven't actually worked in TeX for several years now, so I'm not very familiar with these details anymore. That said, I wonder if there is an easier way to do what you're trying to do. Without looking into exactly how `\Expand` behaves, I would imagine that something like `\Expand{\ifdebug ... \fi}` should already do exactly what you want, shouldn't it?

That's what I first thought as well. `\ifdebug` expands, yes, but just that cs. In effect, I get a dangling `\fi` in the macro (or worse, when I also use `\else`).

```
\newif\ifdebug
\debugtrue

\Inline\def\macro{%
  ...
  \Expand{\ifdebug debug code 2 fi}
  ...
}
```

produces

```
> \macro=macro:
->... debug code 2 fi ... .
```

**Comment by shicks** [comment]
Yes, I can verify that that is exactly what happens, though I'm a little surprised.  My TeX-fu is a little atrophied at this point.

I wrote a regression test into my test suite and tried to figure out how to make this work without extending the package, and I'm stymied.  I tried various combinations of `\UnsafeExpand`, `\unexpanded`, `\expandafter`, etc.  Here's what it looks like is happening.  When we scan the `\Expand`, the `\futurelet` transforms the token stream into

```
\ID@expandnext \Expand { \ifdebug ... \fi } ...
```

When `\ID@expandnext` expands, we end up with

```
\ID@toks \xa \xa \xa { \xa \the \xa \ID@toks \ifdebug ... \fi } \ID@scan ...
```

When `\ID@toks` is trying to execute, it expands the first `\xa` and gets

```
\ID@toks \xa { \the \ID@toks ... \fi } \ID@scan ...
```

I can reproduce this in a very simple example - for whatever reason, expanding an `\if...` in a token register does not gobble the corresponding `\fi`.  Do you by any chance know where this behavior is specified in the TeXbook?  Because it seems pretty surprising.

**Comment by shicks** [comment]
Okay, it's starting to come back to me.  When an `\iftrue` is expanded, it's simply consumed, and nothing else happens.  When an `\iffalse` is expanded, the mouth eats balanced `\if`-`\fi` pairs until it finds either an `\else` or a `\fi` on its own and gobbles the whole thing.  When an `\else` is expanded, it does the same thing, eating balanced `\if`-`\fi` pairs until it finds an unbalanced `\fi` and then gobbles it all.  That's why you often need to write things like `\expandafter\def\fi\whatever` so that the `\fi` in between the `\def` and the `\whatever` disappears before the `\def` executes.

Given this, it seems like the chance of making an `\if` "Just Work™" is slim - we'd need to emulate the balanced parsing (which is tricky with all the ways to spell `\if`: `\iftrue`, `\iffalse`, `\ifnum`, etc), and even then because we only get one shot at expanding, we'd have to consume the `\fi` too early and that would be surprising.  It could possibly do something weird like `\ExpandIf{debug}...\Else...\Fi` and then since we're using our own symbols, it would be easier to parse and slightly less unexpected.

Another possible workaround would be to make something like the following work:

```
\Inline\def\a{\Expand{\ifdebug ... \UnsafeExpand\else ... \UnsafeExpand\fi}}
```

This doesn't seem to get the job done either - for whatever reason the `\else` isn't gobbling the `\fi`.  It's also ugly as all get out.

So I'm starting to believe that a new inline "primitive" is the correct approach.  But I find the `IfT`, `IfF`, `IfTF` style distasteful.  It took me quite a while to figure out `IfT` was talking about "true" when I first read it.  What about the following:

```
\ExpandIf{debug}{ ... }
\ExpandIf{debug}\Else{ ... }
\ExpandIf{debug}{ ... }\Else{ ... }
```

This is pretty readable and analogous to the way we already write `\ifdebug\else ... \fi` for the false-only case, and is pretty economical with additional quarks.  Alternatively it could be `\ExpandIf\debug{ ... }\Else{ ... }`, which saves a character at the cost of a gratuitous and eccentric extra csname-to-string conversion and possible confusion since `\debug` isn't actually a thing.  So I'm inclined to leave the cuteness at `\Else`.

What do you think?

**Comment by sasozivanovic** [comment]
As you say, making an `\if...` work in general seems difficult. Furthermore, some authors will not be using TeX `\if...`' directly but might be fond of `\ifthenelse` from package `ifthen`, or `\ifbool` or `\iftoggle` from package `etoolbox`, or some other mechanism (I'm sure I can't even list all the approaches).

I'd say this argues for the `\NewInlineCommand` approach I suggested in the draft of the extension. This way, each author could define precisely what (s)he needs and how (s)he likes it. After all (and I'm quoting a presupposition from the package documentation here) `inlinedef` provides syntactic sugar. Why not make it as sweet as possible?

I include the relevant definition from the patch I sent (while noting that it should be reworked a bit to define the user commands as quarks - it seems wise to have them defined only in the context of `\Inline`, but I was too lazy to do that in the draft):

```
\newtoks\ID@usercommands
\newcommand\NewInlineCommand[3][0]{%
  \newcommand#2[#1]{#3}%
  \ID@usercommands\xa{%
    \the\ID@usercommands
    \ifx\@foo#2%
      \def\next{\ID@expandunsafe#2}%
    \fi
  }%
}
```

And, of course, `\the\ID@usercommands` should be included somewhere in `\ID@switch`.

Having an `\ExpandIf` (with any syntax, I'm really not fussed about particulars) would of course be handy, but maybe it would be more instructive to include its definition in the documentation, as an example of using `\NewInlineCommand`?

**Comment by sasozivanovic** [comment]
Btw, `\ExpandIf(T)(F)` was inspired by `IfBoolean(T)(F)` from LaTeX3, `xparse` package in particular. I guess they like it. Again, imho, to each his own! ;-)

**Comment by shicks** [comment]
I'm not seeing your proposed definition of `\ExpandIfT` via `\NewInlineCommand` in your patch.  Could you post that as well?  It's not clear to me how you intend it to be used (again, blame my TeX atrophy).

**Comment by sasozivanovic** [comment]
Sure. 

In the original email I put the `NewInlineCommand` in the `.dtx` and the definition of `\ExpandIfT` and friends into a separate file, as an example of usage. Here it is:

```
\documentclass{article}
\usepackage{myinlinedef}

\makeatletter
\def\@ExpandIfTF#1{%
  \csname
    % I'm not 100% sure if this plays well in every situation
    \csname if#1\endcsname
      @firstoftwo%
    \else
      @secondoftwo%
    \fi
  \endcsname
}
\NewInlineCommand[2]\ExpandIfT{%
  \MultiExpand{3}{%
    \@ExpandIfTF{#1}{#2}{}%
  }%
}
\NewInlineCommand[2]\ExpandIfF{%
  \MultiExpand{3}{%
    \@ExpandIfTF{#1}{}{#2}%
  }%
}
\NewInlineCommand[3]\ExpandIfTF{%
  \MultiExpand{3}{%
    \@ExpandIfTF{#1}{#2}{#3}%
  }%
}
\makeatother

\begin{document}

\newif\ifdebug
\debugfalse
\Inline\def\macro{%
  begin code
  \ExpandIfT{debug}{debug code 1 }%
  more code
  \ExpandIfF{debug}{normal code 2 }%
  more code
  \ExpandIfTF{debug}{debug code 3}{normal code 3}
  end code
}
\show\macro


\end{document}
```

**Comment by shicks** [comment]
Thanks for the example.

Your `\csname` trick is pretty slick, but unfortunately it can cause some troubles in rare situations when TeX is skipping over a branch:

```
\def\IF#1#2{\csname\csname if#1\endcsname @gobble\fi\endcsname}
```

This works fine in most cases, but if you write

```
\iffalse
  \def\IF#1#2{\csname\csname if#1\endcsname @gobble\fi\endcsname}
\fi
```

then you're in for a nasty surprise when the unpaired `\fi` in the definition gets in the way.  It's not likely to be a problem, but there's other ways to handle it.  For instance,

```
\def\IF#1#2#3{%
  \begingroup
    \xa\let\xa\X\csname if#1\endcsname
    \xa\ifx\csname iftrue\endcsname\X
      \def\A{#2}%
    \else
      \def\A{#3}%
    \fi
  \xa\endgroup
  \A
}
```

In any case, I'll see what I can put together.

**Comment by shicks** [comment]
I should be able to provide a `\DefineInlineSpecial` macro that acts like `\def` but allows defining commands that get _executed_, rather than simply _expanded_ inside `\Inline`.  This should allow defining all the existing specials this way, which is pretty appealing:

```
\DefineInlineSpecial\NoExpand#1{\Save{#1}}

\DefineInlineSpecial\Super{%
  \ID@checkusage\ID@checkredef{\string\Super}%
  \ifID@aborted\else\xa\xa\xa\Save\xa\xa\xa{\xa\@reservedc\ID@fixedparams}}

\DefineInlineSpecial\Expand#1{\xa\Save\xa{#1}}

\DefineInlineSpecial\MultiExpand#1#2{%
  \def\next{}
  \ifnum#1>\@ne
    \count@#1\advance\count@\m@ne
    \xa\xa\xa\def\xa\xa\xa\next\xa\xa\xa{\xa\xa\xa{\xa\the\xa\count@\xa}\xa{#2}}%
    \xa\def\xa\next\xa{\xa\Continue\xa{\next}}%
  \fi
  \next}

\DefineInlineSpecial\UnsafeExpand{\expandafter}
```

In addition to the low-level `\DefineUnsafeSpecial`, I could imagine providing some higher-level LaTeX-y versions `\NewInlineExpansion` (and maybe even `\RenewInlineExpansion` and `\ProvideInlineExpansion`, though these are probably a bit gratuitous) that take the same sort of arguments as `\newcommand` - specifically, an optional argument count and any optional argument values - and would basically be "unsafe expanded", so that you could write

```
\NewInlineExpansion\Foo[2][\x]{\MultiExpand3{#1{#2}}}
```

and it would do three expansions on whatever was dumped in.

**Comment by sasozivanovic** [comment]
Yes, my definition of `\IF` has the problem you describe. While yours avoids that problem, it has another one: it is not expandable, so it cannot be used inside `\Inline`.

Another way to fix my solution would be

```
\def\IF#1#2{\csname\csname if#1\endcsname @gobble\csname fi\endcsname\endcsname}
```

**Comment by sasozivanovic** [comment]
Ahh, I see now that you have actually addressed the nonexpandability problem by `\DefineInlineSpecial`. I'll have to study it a bit more but the idea seems nice. (Another one that might be useful is `\ExpandFully`.)
